### ISSUE TITLE
Session map entries not removed when session expires

### ISSUE THREAD

**Issue by MattNohelty** [hazelcast/hazelcast-wm#48]
Hazelcast is not removing destroyed sessions from the underlying map used for session clustering.  I'm using Hazelcast 3.8 with WebFilter and SessionListener for session clustering.

WebFilter is configured to connect to an external Hazelcast cluster and store sessions in a map. Here's a subset of the web.xml filter configuration.

```
<init-param>
    <param-name>use-client</param-name>
    <param-value>true</param-value>
 </init-param>
 <init-param>
    <param-name>map-name</param-name>
    <param-value>SessionMap</param-value>
 </init-param>
```

In web.xml, my session-timeout is set to 120 (i.e. 2 hours). The Hazelcast map, named SessionMap, has a TTL of 24 hours (configured on the external cluster, not in web.xml). The thinking here is that we want to end sessions that are inactive for 2 hours but if someone is continuously using the application we want to let the session continue for a max of 24 hours.

The problem is that if I leave a session idle, Tomcat will terminate the session after 2 hours but the session will still remain in the Hazelcast map (SessionMap) until the TTL on that map.  

When a session is destroyed by my application server (Tomcat), SessionListener calls WebFilter#destroyOriginalSession.  This calls WebFilter#destroySession with the invalidated argument set to false.  See below:

```
void destroyOriginalSession(HttpSession originalSession) {
    String hazelcastSessionId = (String)this.originalSessions.remove(originalSession.getId());
    if(hazelcastSessionId != null) {
        HazelcastHttpSession hazelSession = (HazelcastHttpSession)this.sessions.get(hazelcastSessionId);
        if(hazelSession != null) {
            this.destroySession(hazelSession, false);
        }
    }
}
```

This eventually creates a DeleteSessionEntryProcessor which is executed with executeOnKey().  The code from DeleteSessionEntryProcessor is below.  “this.invalidate” is always false because of the above code, so it never actually removes the entry from the map.  

```
public Object process(Entry<String, SessionState> entry) {
    SessionState sessionState = (SessionState)entry.getValue();
    if(sessionState == null) {
        return Boolean.FALSE;
    } else {
        if(this.invalidate) {
            entry.setValue((Object)null);
            this.removed = true;
        } else {
            entry.setValue(sessionState);
        }

        return Boolean.TRUE;
    }
}
```

I modified WebFilter#destroyOriginalSession and passed in true and it seems to behave correctly and remove the entries from the underlying map as I would expect.  

As a side note, I opened a Stack Overflow question this morning before I dug into this further: http://stackoverflow.com/questions/43217697/hazelcast-remove-expired-session-from-map

**Comment by MattNohelty** [comment]
I stumbled upon an old bug from 2014 (https://github.com/hazelcast/hazelcast/issues/2930) which is reporting the same behavior.  After reading that bug description, I realize the solution proposed above is not sufficient as it would prematurely remove the map entries when any session is destroyed even if that session still exists on other servers.  

However, it looks like this issue was resolved by bturner (https://github.com/bturner/hazelcast/commit/21c679e8a1c8226ecb3d7597b48a6bbd2d011a52) in version 3.3 by keeping track of the number of remaining sessions and removing the session map entry if there are no remaining original sessions. 

Was this fix removed for some reason?

**Comment by dfex55** [comment]
Behavior confirmed using hazelcast 3.8 and hazelcast-wm 3.8 on jetty.

web.xml:
```
  <session-config>
    <session-timeout>2</session-timeout>
    <cookie-config>
      <name>jetty.id</name>
      <path>/</path>
    </cookie-config>
  </session-config>
  <listener>
    <listener-class>com.hazelcast.web.SessionListener</listener-class>
  </listener>
```

hazelcast:
```
sticky-session = false
deferred-write = true
```
Does not remove hazelcast map entry after two minutes.

But not sure if invalidating hazelcast session with sticky-session = false configured should happen at all. Assume you visited one node in the WebContainer-Cluster, but never revisit it again (loadbalancer decision), the node should not invalidate the hazelcast session (through SessionListener and local session timeout), since the session might be used actively by other nodes.

**Comment by MattNohelty** [comment]
@dfex55  - I agree with you that you cannot blindly remove the map entry when `sticky-session = false`.  However, the fix by bturner here https://github.com/bturner/hazelcast/commit/21c679e8a1c8226ecb3d7597b48a6bbd2d011a52 seems to address that exact case by keeping track of how many nodes have that session and only removing it from the map when there are no nodes holding that session.  

At the very least, it seems like the session map entry should be removed when `sticky-session = true`.

**Comment by mesutcelik** [comment]
Hi @MattNohelty 

I remember there were some problems with counter based solutions.
* Imagine that you are using sticky=true and you only have one active session in one tomcat container among a cluster of tomcat containers. If that container is restarted or app redeployed, `sessionDestroy` is kicked in and as counter = 1 then your session is removed from cluster even though you still have active session
* in case of a tomcat crash, counter is not decremented which might cause some memory issues over time.

In the new design, you are somehow obliged to configure the Hazelcast TTL separate from tomcat configuration which should be equal to each others in normal cases but you can give longer TTL in hazelcast.xml if you want to continue to use the cached session data even though users logs in even though his session is expired already...

I am probably missing something but i don't get why the current implementation does not cover your case. If you configure TTL = 24 hours then your distributed session will stay in IMap for 24 hours even after your session is expired after 2 hours in the Tomcat container. You can reuse session as long as your cookie contains the same `hazelcast.sessionId` and your user comes in before 24 hours.

Can you please explain a bit in more details how you are using 2 hours and 24 hours concepts in your app then we can be in the same page?

**Comment by MattNohelty** [comment]
We would like our sessions to time out after 2 hours of inactivity but never last longer than 24 hours, regardless of activity.  Here are some sample use cases:
- A user authenticates and then leaves their desk for 3 hours.  When they come back, they should have to authenticate again.
- A user who works in the application for 8 hours straight should never be prompted to log in while working. 
- A user who works in the application for 30 hours straight should be prompted to log back in after 24 hours.

Our relevant settings are as follows:
- Tomcat session timeout is 2 hours.
- Hazelcast session map max idle time is 2 hours.
- Hazelcast session map TTL is 8 hours.

With these settings, a user who is inactive for 2 hours should have their session invalidated in Tomcat and in the Hazelcast map (because of idle time).

I'm have a test system that I'm working on now with settings like this:
- Tomcat session timeout is 1 minute.
- Hazelcast session map max idle time is 1 minute.
- Hazelcast session map TTL is 10 minutes.

I'm able to reproduce this same behavior where sessions appear to get stuck in the map but I can't reproduce consistently so I'm trying to nail down what condition is causing the difference in behavior.  If I submit a bunch of authentication requests and then wait for longer than 10 minutes, I would expect the map to be completely empty but it's not.

**Comment by MattNohelty** [comment]
I've still not been able to determine a pattern for when this happens but I added a Map Entry Listener on the session map and certain entries appear to be ignored by the idle time and TTL settings.  I see the session entry get added/updated when I authenticate but the entryEvicted handler (or any other handler method) is never called after the idle time or the TTL so the entry remains in the map.  Again, this behavior is not consistently reproducible so sometimes I do see the entry get evicted at the idle time.

**Comment by MattNohelty** [comment]
@mesutcelik - Can you define exactly how idle time and TTL are supposed to work with regards to the session map?  

When an app server session expires, the DeleteSessionEntryProcessor is called with `invalidate` set to false so `entry.setValue(sessionState);` is executed.  Based on what I'm seeing in my map listener, this updates the session map so it presumably resets the idle timer.  This doesn't seem right.  I would only expect the idle time to be reset if the user actually does something, not when the app server session times out.  If a user stops using the application, I would expect their session to be terminated after the max idle time passes.

Also, I found this StackOverflow post from late 2015 (http://stackoverflow.com/questions/33352262/ttl-value-of-an-entry-in-map-is-changed-for-every-update-of-that-entry) which seems to indicate that the TTL gets reset after each update to the map.  This also doesn't seem right.  If I have a TTL of 10 minutes on my session map, I would expect no session to last longer than 10 minutes, regardless of the user's activity.  

The crux of my issue is that our application has a large number of sessions and some of them do not seem to ever be getting removed from our session map which is causing memory issues. Most of our users don't explicitly log out so we need to be able to rely on some combination of app server session timeout, session map idle time, or session map TTL to ensure that our old sessions are always getting cleaned up.  

On my test system (1 minute app server session timeout, 1 minute session map max idle time, and 10 minute session map TTL), I have a session that's been unused for over an hour and it's still in the map.  The app server session timed out after 1 minute as expected.  This triggered a map entry update (as described above) but there have been no other map entry listeners called since then.  Both the idle time and TTL should have expired many times over.

To help debug this, is there a good way to access the idle and TTL timers for a map entry so I can see what's happening with this particular session?

**Comment by MattNohelty** [comment]
I was able to debug into Hazelcast and feel pretty confident that my TTL and max idle time configurations are being picked up and set correctly but I still have the questions from my previous post about how exactly these are supposed to work.

One interesting thing I noticed is that map entry eviction seems to stop.  From what I can tell, the process to detect map evictions begins in ClearExpiredOperation.java.  The run method loops over all maps and will further inspect the entries in the map for eviction if the map’s size is greater than zero and the map is expirable.  

```
@Override
    public void run() throws Exception {
        final MapService mapService = getService();
        MapServiceContext mapServiceContext = mapService.getMapServiceContext();
        final PartitionContainer partitionContainer = mapServiceContext.getPartitionContainer(getPartitionId());
        final ConcurrentMap<String, RecordStore> recordStores = partitionContainer.getMaps();
        final boolean backup = !isOwner();
        for (final RecordStore recordStore : recordStores.values()) {
            if (recordStore.size() > 0 && recordStore.isExpirable()) {
                recordStore.evictExpiredEntries(expirationPercentage, backup);
                recordStore.disposeDeferredBlocks();
            }
        }
    }
```

When everything is working, `recordStore.size() > 0` is true, and after the idle time passes, map entries get evicted by AbstractEvictableRecordStore#getOrNullIfExpired.  

After I use the application for a while (create a few sessions, let some sessions expire, etc.), eventually `recordStore.size()` always returns 0 and from that point forward my session map entries never get evicted. This happens even when there are things in the actual map.  I confirmed this by running HazelcastInstance.getMap("SessionMap") on the client and there are sessions in the map.

I haven't been able to determine what exactly triggers this scenario so I cannot reproduce it consistently.  I'll continue to dig into this but am not at all familiar with the Hazelcast code base so any guidance you could provide would be greatly appreciated.  Is anything I said above incorrect?  Have you received any other bug reports similar to this?

**Comment by emre-aydin** [comment]
> To help debug this, is there a good way to access the idle and TTL timers for a map entry so I can see what's happening with this particular session?

[Accessing Entry Statistics section](http://docs.hazelcast.org/docs/3.8/manual/html-single/index.html#accessing-entry-statistics) on docs might help.

**Comment by emre-aydin** [comment]
> On my test system (1 minute app server session timeout, 1 minute session map max idle time, and 10 minute session map TTL), I have a session that's been unused for over an hour and it's still in the map. 

If the map that's holding the sessions have a TTL of  10 minutes, there should be no way that session is still on the distributed map. There needs to be another code path (either from within `WebFilter` or your application code) that inserts that same key into the map after it is evicted from the map. 

For debugging, you can take advantage of [MapListeners](http://docs.hazelcast.org/docs/3.8/manual/html-single/index.html#listening-to-map-entries-with-predicates), particularly [EntryExpiredListener](https://github.com/hazelcast/hazelcast/blob/b1c425e8598cde123ab11493b4dc7896dbfec15e/hazelcast/src/main/java/com/hazelcast/map/listener/EntryExpiredListener.java) might be of help.

**Comment by MattNohelty** [comment]
@emre-aydin - Thank you for pointing me to the Entry Statistics.  This was very helpful.  The issue seems to be that the expiration time is being set inconsistently when new records are added.  Here's the output for a few of my tests this morning:

My test system is similar to what I mentioned in the previous post except I changed the Hazelcast Idle time setting to be slightly lower than the session time out so they wouldn’t fire at the same time. (1 minute app server session timeout, 50 second session map max idle time, and 10 minute session map TTL).

Test 1: Fully stop test application and then restart so everything in Hazelcast starts from a fresh state.  I create the first session and it has an expirationTime of Long.MAX_VALUE which is not right.  As you would expect, this map entry never gets evicted. 

```
Time: 1491836028039
size in memory  : 1305
creationTime    : 1491836022183
expirationTime  : 9223372036854775807
number of hits  : 2
lastAccessedTime: 1491836022214
lastUpdateTime  : 1491836022214
version         : 1
key             : HZD1802443B8C44C8DAFAA6575C9702236
value           : SessionState {…}
```

Test 2: This is an exact repeat of Test 1.  Stop and restart everything and then I create the first session and this time it has an expirationTime 50,000 milliseconds after the creation time which is what I would expect (because of the 50 second idle time) and this time the session gets evicted from the map as expected.

```
Time: 1491836551138
size in memory  : 1305
creationTime    : 1491836549793
expirationTime  : 1491836599813
number of hits  : 2
lastAccessedTime: 1491836549813
lastUpdateTime  : 1491836549813
version         : 1
key             : HZC78233BE41E2435EAE9C74BCE6B485FD
value           : SessionState {…}
```

Continuing on with Test 2, I create a few more sessions.  The next one I create has it’s expirationTime set correctly but the following one is set to Long.MAX_VALUE.  

```
Time: 1491837181755
size in memory  : 1305
creationTime    : 1491837152125
expirationTime  : 1491837202130
number of hits  : 2
lastAccessedTime: 1491837152130
lastUpdateTime  : 1491837152130
version         : 1
key             : HZ0FF2CE9E0F9E4DC593E97918C8B757DC
value           : SessionState {…}

size in memory  : 1305
creationTime    : 1491837179270
expirationTime  : 9223372036854775807
number of hits  : 2
lastAccessedTime: 1491837179286
lastUpdateTime  : 1491837179286
version         : 1
key             : HZ9DAAA32A954F471099FA6952CEA5033A
value           : SessionState {…}
```

So it seems like the root cause of the problem is that the expiration time is getting set incorrectly for some of these session map entries.  Have you seen this before?  Any ideas for what could cause this or how to track it down?  As far as I'm aware, there is a single config for a given map across the entire cluster so I don't understand what would cause this inconsistency.

**Comment by MattNohelty** [comment]
It appears that the nodes are using a different MapConfig for the session map.  One node is using the predefined map config with max idle time and TTL and the other node is using a map config called “default” which does not have a max idle time or TTL.   I have been operating under the assumption that each map can have one and only one configuration across the entire cluster.  Is this not correct?

We are using programmatic configuration to set our map configuration.  When each node starts up, it loops through all of our predefined maps and does the following:

```
Config config = hazelcastInstance.getConfig();
config.getMapConfig(“sessionMap”).setTimeToLiveSeconds(timeToLiveSeconds);
config.getMapConfig(“sessionMap”).setMaxIdleSeconds(idleSeconds);
```

The above is run at startup on each node so the map config is defined multiple times.  My understanding is that the last configuration set is the one that take effect across the entire cluster. However, this really shouldn’t matter because the same configuration is used by each node.  

I created a background thread that runs on each node that prints out the configuration for the session map every few seconds (see below). This prints out the correct configuration on all nodes (i.e. the same configuration that was specified in the code above)

```
MapConfig mapConfig = hazelcastInstance.getConfig().getMapConfig("sessionMap ");
logger.debug("Map Config (Name): " + mapConfig.getName());
logger.debug("Map Config (TTL): " + mapConfig.getTimeToLiveSeconds());
logger.debug("Map Config (Idle): " + mapConfig.getMaxIdleSeconds());
```

However, if I set breakpoints in the Hazelcast code (e.g. `MapOperation#innerBeforeRun`) the `mapContainer.getMapConfig()` for “sessionMap” is set correctly on one node and set to “default” on the other and it looks like the entries that have expirationTime set to Long.MAX_VALUE are happening when they get created on the node that is using the “default” config.

Are any of the assumptions I mentioned above incorrect?  Is there anything wrong with the configuration process I described?  I'm almost positive this is the root cause of our issue so now I need some help understanding how/why it gets into this state.

**Comment by emre-aydin** [comment]
@MattNohelty 

You need to provide the configuration when starting the Hazelcast instance. See the [docs for an example](http://docs.hazelcast.org/docs/3.8/manual/html-single/index.html#configuring-programmatically) on how to do it. 

Getting the config object of an already started cluster and changing settings there has no effect. Following warning is from the docs:
> NOTE: The Config must not be modified after the Hazelcast instance is started. In other words, all configuration must be completed before creating the HazelcastInstance.

**Comment by MattNohelty** [comment]
@emre-aydin - You are right.  That seems to be our issue.  Changing the configuration after the fact did work in some cases but definitely not consistently.  Thanks for your help.  This issue can be closed.

**Comment by emre-aydin** [comment]
@MattNohelty Glad to hear your problem is resolved.

**Comment by yifengwu17** [comment]
@MattNohelty I encountered the same problem as you. How did you solve it in the end?
